python/ray/remote_function.py

```diff
@@ -36,14 +27,14 @@ def compute_function_id(function):
     # Include the function module and name in the hash.
     function_id_hash.update(function.__module__.encode("ascii"))
     function_id_hash.update(function.__name__.encode("ascii"))
-    # If we are running a script or are in IPython, include the source code in
-    # the hash. If we are in a regular Python interpreter we skip this part
-    # because the source code is not accessible. If the function is a built-in
-    # (e.g., Cython), the source code is not accessible.
-    import __main__ as main
-    if (hasattr(main, "__file__") or in_ipython()) \
-            and inspect.isfunction(function):
-        function_id_hash.update(inspect.getsource(function).encode("ascii"))
+    try:
+        # If we are running a script or are in IPython, include the source code
+        # in the hash.
+        source = inspect.getsource(function).encode("ascii")
+        function_id_hash.update(source)
+    except OSError:
```

Comment:
This doesn't seem to work for me in a Python 2 interpreter. It raises an `IOError`, which is related, but slightly different.

```
22:05 $ python2
Python 2.7.13 |Anaconda 4.3.1 (x86_64)| (default, Dec 20 2016, 23:05:08) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import inspect
>>> def f():
...     pass
... 
>>> try:
...     inspect.getsource(f)
... except OSError as e:
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/Users/rkn/anaconda2/lib/python2.7/inspect.py", line 700, in getsource
    lines, lnum = getsourcelines(object)
  File "/Users/rkn/anaconda2/lib/python2.7/inspect.py", line 689, in getsourcelines
    lines, lnum = findsource(object)
  File "/Users/rkn/anaconda2/lib/python2.7/inspect.py", line 537, in findsource
    raise IOError('could not get source code')
IOError: could not get source code
```

Reply:
Ah yes, this changed in Python 3.3 it seems: https://docs.python.org/3/library/inspect.html

I'll change it to catch both for compatibility.